### PR TITLE
DMP-2533: Implement PATCH /admin/retention-policy-types

### PR DESCRIPTION
**Pull request by chrisbellingham-hmcts**
## [DMP-2533](https://tools.hmcts.net/jira/browse/DMP-2533)

Implement new endpoint `PATCH /admin/retention-policy-types/<id>`

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/darts/retention/service/impl/RetentionPolicyTypeServiceImpl.java [hunk 1]
```diff
@@ -78,4 +97,108 @@ public RetentionPolicyType createOrReviseRetentionPolicyType(AdminPostRetentionR
...
+
+        final String effectiveFixedPolicyKey = suppliedFixedPolicyKey == null ? originalFixedPolicyKey : suppliedFixedPolicyKey;
+        final OffsetDateTime suppliedPolicyStartAt = adminPatchRetentionRequest.getPolicyStartAt();
+        if (suppliedPolicyStartAt != null && !suppliedPolicyStartAt.isEqual(targetEntity.getPolicyStart())) {
+            handlePolicyStartAtChange(effectiveFixedPolicyKey, suppliedPolicyStartAt, targetEntity);
+        }
+
+        String suppliedName = adminPatchRetentionRequest.getName();
+        if (suppliedName != null) {
+            policyNameIsUniqueValidator.validate(adminPatchRetentionRequest.getName(), effectiveFixedPolicyKey);
```
[the remark this answers is not in the document]
Comment (chrisbellingham-hmcts) [reply src/main/java/uk/gov/hmcts/darts/retention/service/impl/RetentionPolicyTypeServiceImpl.java:121]: Yep, will update.

File: src/main/java/uk/gov/hmcts/darts/retention/service/impl/RetentionPolicyTypeServiceImpl.java [hunk 2]
```diff
@@ -78,4 +97,108 @@ public RetentionPolicyType createOrReviseRetentionPolicyType(AdminPostRetentionR
...
+            targetEntity.setPolicyName(suppliedName);
+        }
+
+        String suppliedDisplayName = adminPatchRetentionRequest.getDisplayName();
+        if (suppliedDisplayName != null) {
+            policyDisplayNameIsUniqueValidator.validate(adminPatchRetentionRequest.getDisplayName(), effectiveFixedPolicyKey);
```
[the remark this answers is not in the document]
Comment (chrisbellingham-hmcts) [reply src/main/java/uk/gov/hmcts/darts/retention/service/impl/RetentionPolicyTypeServiceImpl.java:127]: Yep, will update.

File: src/main/java/uk/gov/hmcts/darts/retention/service/impl/RetentionPolicyTypeServiceImpl.java [hunk 3]
```diff
@@ -78,4 +97,108 @@ public RetentionPolicyType createOrReviseRetentionPolicyType(AdminPostRetentionR
...
+        }
+
+        String suppliedDescription = adminPatchRetentionRequest.getDescription();
+        if (suppliedDescription != null) {
+            targetEntity.setDescription(suppliedDescription);
+        }
+
+        String suppliedDuration = adminPatchRetentionRequest.getDuration();
+        if (suppliedDuration != null) {
+            policyDurationValidator.validate(adminPatchRetentionRequest.getDuration());
```
[the remark this answers is not in the document]
Comment (chrisbellingham-hmcts) [reply src/main/java/uk/gov/hmcts/darts/retention/service/impl/RetentionPolicyTypeServiceImpl.java:138]: Yep, will update.

File: src/main/java/uk/gov/hmcts/darts/retention/validation/PolicyNameIsUniqueValidator.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+import uk.gov.hmcts.darts.retention.exception.RetentionApiError;
+
+@Component
+@RequiredArgsConstructor
+public class PolicyNameIsUniqueValidator implements BiValidator<String, String> {
+
+    private final RetentionPolicyTypeRepository retentionPolicyTypeRepository;
+
+    @Override
+    public void validate(String name, String fixedPolicyKey) {
```
[the remark this answers is not in the document]
Comment (chrisbellingham-hmcts) [reply src/main/java/uk/gov/hmcts/darts/retention/validation/PolicyNameIsUniqueValidator.java:17]: Sure, will rename `name` to `policyName`

**Review by johnrace0 — APPROVED** [review]

**Pull request merged**